Add optional feature flag param to get domain membership info

akashkj opened the pull request:
@proteusvacuum @sravfeyn @kaapstorm 

## Product Description
Enhanced the API endpoint for getting  domain membership info for a user by providing additional filter for feature flag.

## Technical Summary
https://dimagi-dev.atlassian.net/browse/SC-2069

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
Performed local testing

### Automated test coverage
TODO: Add unit test


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

AmitPhulera committed:
fix failing tests

proteusvacuum committed:
Add UCRExpression model

proteusvacuum committed:
Add management view for UCRExpression

proteusvacuum committed:
Use dataclass for FactoryContext

AmitPhulera committed:
update: object creation logic

AmitPhulera committed:
Merge branch 'ap/sql-case-repeater' of https://github.com/dimagi/commcare-hq into ap/sql-case-repeater

AmitPhulera committed:
add: tests to verify __new__ works correctly

AmitPhulera committed:
add: sync fns to the cowin repeaters

AmitPhulera committed:
add: migration command for cowin repeaters

AmitPhulera committed:
add: cowin repeaters to parent migration cmd

AmitPhulera committed:
lint

AmitPhulera committed:
update: migration file

proteusvacuum committed:
Use UCR_EXPRESSINO_REGISTRY feature flag

proteusvacuum committed:
Update FactoryContext to pull filters and expressions from DB

proteusvacuum committed:
Update references to FactoryContext to include domain

proteusvacuum committed:
Add UCRExpression to modeldeletions

proteusvacuum committed:
Add description field to UCRExpression

proteusvacuum committed:
Add description to UCRExpression views

proteusvacuum committed:
Use model manager for wrapped definitions

proteusvacuum committed:
Pull FactoryContext creation out of loops

proteusvacuum committed:
Use dict union operator

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into fr/ucr-registry-linked

proteusvacuum committed:
Use django 3 to make migration

proteusvacuum committed:
fix bad merge

snopoke committed:
remove jad / jar view

snopoke committed:
remove model functions

snopoke committed:
update views

snopoke committed:
remove tests

snopoke committed:
remove download option

stephherbers committed:
pass through isMultiSelect variable insetead of hardcoded flag

stephherbers committed:
clear sessionStorage.selectedValues once selectedValues is set

gherceg committed:
Merge branch 'gh/erm/remove-linked-domain-ff' into gh/erm/remove-linked-domain-ff+fr/ucr-registry-linked

gherceg committed:
Merge branch 'gh/erm/remove-linked-domain-ff' into gh/erm/remove-linked-domain-ff+fr/ucr-registry-linked

gherceg committed:
Merge branch 'fr/ucr-registry-linked' into gh/erm/remove-linked-domain-ff+fr/ucr-registry-linked

gherceg committed:
Merge branch 'gh/erm/remove-linked-domain-ff' into gh/erm/remove-linked-domain-ff+fr/ucr-registry-linked

AmitPhulera committed:
add: RepeaterMeta class for automatic repeater class map population

AmitPhulera committed:
remove: REPEATER_CLASS_MAP of util

AmitPhulera committed:
add: metaclass in sqlrepeater

AmitPhulera committed:
update: error message

AmitPhulera committed:
update: move only exception prone code in try block

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ap/sql-case-repeater

AmitPhulera committed:
update tests

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

AmitPhulera committed:
Merge branch 'ap/sql-case-repeater' of https://github.com/dimagi/commcare-hq into ap/sql-case-repeater

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into fr/ucr-registry-linked

joxl committed:
Add design details to README.rst

joxl committed:
Improve multiplexing description in README.rst

Co-authored-by: Simon Kelly <skelly@dimagi.com>

biyeun committed:
create endpoint to submit hubspot cta form from the backend

biyeun committed:
add submit_hubspot_cta_form as registered url to base.html

biyeun committed:
recreate hubspot CTA forms as a native form

biyeun committed:
replace hubspot CTA forms with the new native knockout forms since adblockers block hubspot scripts and this is a critical workflow

biyeun committed:
make get_client_ip_from_meta a utility

biyeun committed:
actually submit hubspot CTA form from submit_hubspot_cta_form endpoint

biyeun committed:
make log_response a general analytics utility and log the response of the backend hubspot CTA form submission

joxl committed:
Outline "primary/secondary swap" step as intended part of the procedure.

joxl committed:
Merge branch 'master' into jm/elastic-client-adapter

joxl committed:
Prevent regression from settings updates

- `case_search` and `forms` support custom index names via
  localsettings.py

biyeun committed:
small fix, remove the  as it doesn't make sense when an email comes after it

biyeun committed:
prepare to split the SSOEnterpriseSettingsForm for SAML and OIDC

stephherbers committed:
move isMultiSelect declaration
master rebase conflicts

joxl committed:
Merge branch 'master' into jm/elastic-client-adapter

stephherbers committed:
replace punctuation

biyeun committed:
move entity_id to base form

biyeun committed:
make service_name a property of the IdentityProvider and make form text less specific to Azure AD

biyeun committed:
add abstract method to update identity provider to the base form

biyeun committed:
add enterprise settings form for oidc

biyeun committed:
add tests for SsoOidcEnterpriseSettingsForm

stephherbers committed:
deselect select all checkbox if an item is deselected from list

AmitPhulera committed:
add: populating repeater_class_map logic in repeaters apps.py

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ap/sql-case-repeater

AmitPhulera committed:
update: migration file

snopoke committed:
replace URL selections with whatever Formplayer returns

snopoke committed:
remove key from sessionStorage completely

orangejenny committed:
Added toggle utility functions

orangejenny committed:
Updated build_errors rendering to accept full list of toggles

orangejenny committed:
Fixed test

Test test_download_file_bad_xform_404 uses an AnonymousUser, which doesn't have an email attribute.

orangejenny committed:
Added cache invalidation and updated naming

orangejenny committed:
Added email toggles to SessionDetailsView

orangejenny committed:
Updated util functions and callers

orangejenny committed:
Revert "Updated util functions and callers"

This reverts commit cf1a7263477b5cac21f879e1f087ad89d884ec1b.

orangejenny committed:
Revert "Added email toggles to SessionDetailsView"

This reverts commit f35ced54e800f90c69684a546a41e1307f03778d.

dependabot[bot] committed:
Bump grunt from 1.4.1 to 1.5.2

Bumps [grunt](https://github.com/gruntjs/grunt) from 1.4.1 to 1.5.2.
- [Release notes](https://github.com/gruntjs/grunt/releases)
- [Changelog](https://github.com/gruntjs/grunt/blob/main/CHANGELOG)
- [Commits](https://github.com/gruntjs/grunt/compare/v1.4.1...v1.5.2)

---
updated-dependencies:
- dependency-name: grunt
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] <support@github.com>

AmitPhulera committed:
add check for deferred fields on new

AmitPhulera committed:
remove: meta class for repeaters as repeater_map is populated in app config

orangejenny committed:
Removed unused id

orangejenny committed:
Removed unused event

orangejenny committed:
Moved case selection handling into CaseDetailFooterView

AmitPhulera committed:
remove: unused import:

AmitPhulera committed:
feedback from PR

orangejenny committed:
Added triggers for multiSelect events

orangejenny committed:
Combined multi-select events into single event

Writing both add and remove handlers was going to get annoying.

orangejenny committed:
Added handler to check row box

orangejenny committed:
Moved reconcileSelectAll into event handler

orangejenny committed:
Added updating of selectedValues into event handler

orangejenny committed:
Removed TODO

orangejenny committed:
Removed selectedCaseIds updating from selectAllAction

orangejenny committed:
Moved updateContinueButtonText logic into event handler

orangejenny committed:
Removed selectedValue updating from selectRowAction

This is now handled in the CaseListView event handler.

orangejenny committed:
Switched parentView option to isMultiSelect in CaseView

orangejenny committed:
Updated CaseListView to share UI reconciling code between event handler and onRender

orangejenny committed:
Updated interactions with sessionStorage

This is a bit messy. The general idea is to clear selected values whenever the CloudcareURL
would clear selections. This PR also moves some of the UI-updating code into onRender to guarantee that the
elements in ui are available.

orangejenny committed:
Merge branch 'smh/multi-select-qa-followup' into smh+jls/multi-select-select-button

orangejenny committed:
Merge branch 'smh/multi-select-case-detail-workflow' into smh+jls/multi-select-select-button

kaapstorm committed:
Fix parameter order

Commit:
Add optional feature flag param to get domain membership info

Commit:
Add optional feature flag param to get domain membership info

snopoke reviewed (commented):
Can you give more information about the use case this. The way it has been implemented seems very weird. It will only return domains that have the FF enabled.

orangejenny committed:
Merge branch 'smh/multi-select-clear-session-storage' into jls/multi-select-select-button

orangejenny committed:
Update staging.yml

orangejenny committed:
Removed removal from app:select:menus

This was causing QA-4110, clearing the selections when the user changes pages.

orangejenny committed:
Merge pull request #31532 from dimagi/jls/toggles-in-build-errors

Make toggles available to build errors

orangejenny committed:
Update staging.yml

orangejenny committed:
Updated UI elements to use jQuery instead of direct DOM elements

The jQuery style is more forgiving when elements don't exist.

snopoke committed:
Revert "use real types instead of namedtuples"

This reverts commit 9c8ec2ee8715d45197d0a0b8f929505a069c58b2.

snopoke committed:
Merge pull request #31555 from dimagi/sk/revert-type-change

Revert "use real types instead of namedtuples"

AddisonDunn committed:
Update translations.

[ci skip]

biyeun committed:
fix test

biyeun committed:
Update staging.yml

esoergel committed:
Add some tests demonstrating current behavior

biyeun committed:
Merge pull request #31547 from dimagi/dependabot/npm_and_yarn/grunt-1.5.2

Bump grunt from 1.4.1 to 1.5.2

minhaminha committed:
Update staging.yml

orangejenny committed:
Updated multi select select parent first logic

orangejenny committed:
Update staging.yml

orangejenny committed:
Fixed conditional

snopoke committed:
use correct argument name

TypeError: __init__() got an unexpected keyword argument 'status_code'

snopoke committed:
other instances of status_code -> status

snopoke committed:
Merge pull request #31558 from dimagi/sk/status-code

use correct argument name

orangejenny committed:
Merge pull request #31557 from dimagi/jls/rework-multiple-referrals-spf

Updated multi select select parent first logic

AmitPhulera committed:
Merge pull request #30348 from dimagi/ap/sql-case-repeater

[couch-to-sql] [Repeaters] Migrate Repeaters to SQL

orangejenny committed:
Merge pull request #31551 from dimagi/jls/multi-select-select-button

Multi-select case detail "continue" flow

esoergel committed:
Update lazy translation docs

esoergel committed:
Merge pull request #31559 from dimagi/es/multiselect-datums

Add some tests demonstrating current behavior

esoergel committed:
Merge pull request #31560 from dimagi/es/translation-docs

Update lazy translation docs

AddisonDunn committed:
modify SelectToggle to use JSONEncoder so we can pass gettext_lazy text choices

and fix spacing

AddisonDunn committed:
modify SelectToggle choices in various places to use lazy translation

orangejenny committed:
Updated get_add_case_preloads_case_id_xpath to check for multi-select

orangejenny committed:
Update staging.yml

orangejenny committed:
Merge pull request #31562 from dimagi/jls/multi-select-xform-datum

Updated get_add_case_preloads_case_id_xpath to check for multi-select

proteusvacuum committed:
Include context

proteusvacuum committed:
updating files: scripts/staging.yml

[ci skip]

kaapstorm committed:
Merge pull request #31553 from dimagi/nh/param_order

Fix parameter order

AddisonDunn committed:
Merge pull request #31561 from dimagi/ad/SelectToggle-to-gettext_lazy

Allow gettext_lazy with the SelectToggle

orangejenny committed:
Added DetailTab.nodeset_filter

orangejenny committed:
Updated suite generation

orangejenny committed:
Added test

orangejenny committed:
Added UI

calellowitz committed:
fix redis check for multiple hosts

biyeun committed:
Update corehq/apps/analytics/views.py

Co-authored-by: Joel M <jmiller@dimagi.com>

biyeun committed:
Update corehq/apps/analytics/views.py

Co-authored-by: Joel M <jmiller@dimagi.com>

biyeun committed:
Update corehq/apps/sso/forms.py

Co-authored-by: Joel M <jmiller@dimagi.com>

biyeun committed:
Update corehq/apps/sso/forms.py

Co-authored-by: Graham Herceg <gherceg@dimagi.com>

orangejenny committed:
Moved HTML into HTML file

biyeun committed:
fix syntax errors from suggestion

biyeun committed:
Merge pull request #31515 from dimagi/bmb/trial-signup-bug

Fix trial signup form when ad blocker is enabled

biyeun committed:
Update staging.yml

orangejenny committed:
Update staging.yml

calellowitz committed:
Merge pull request #31564 from dimagi/ce/redis-check

fix redis check for multiple hosts

orangejenny committed:
Update staging.yml

Commit:
add aris labels to inputs and add tabindex to heading, field names

Robert-Costello committed:
Update staging.yml

Commit:
Increase Limit of number of upload per block and its enhancement

esoergel committed:
Remove custom xmlns from bulk form updates

https://docs.google.com/document/d/1t97k35y9juT7Wl_suOiWi-NYP3cpiaV4hUmZ0j2e-e4/edit#
We are not consistent in how we use these meta properties with system
forms, but this is more in line with the docstring on submit_case_blocks

biyeun committed:
Merge pull request #31527 from dimagi/bmb/sso-update-enterprise-admin-views

[SSO] Update IdP Enterprise Admin Views to support OIDC

proteusvacuum committed:
Add url template to configurable repeaters

Robert-Costello committed:
Update staging.yml

proteusvacuum committed:
Add url template to form

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into fr/ucr-registry-linked

Commit:
Merge branch 'master' into rc/case-search-accessibility
fix conflict

Robert-Costello committed:
Update staging.yml

gherceg committed:
setup django app registry prior to building docs

gherceg committed:
ignore stickler

Co-authored-by: Joel M <jmiller@dimagi.com>

esoergel committed:
xmlns can accept a list

Commit:
Merge pull request #31569 from dimagi/aj-samveg-per-block-upload

Increase Limit of number of upload per block and its enhancement

Robert-Costello committed:
Update staging.yml

Robert-Costello committed:
Update staging.yml

calellowitz committed:
remove unnecessary string formatting

calellowitz committed:
only write to redis if throttling

calellowitz committed:
Merge pull request #31573 from dimagi/ce/redis-check

remove unnecessary string formatting

gherceg committed:
Merge pull request #31571 from dimagi/gh/fix-docs-build

setup django app registry prior to building docs

esoergel committed:
Merge pull request #31572 from dimagi/es/system-xmlns

Don't set custom xmlns in bulk case util

Robert-Costello committed:
Update staging.yml

gherceg committed:
Revert "setup django app registry prior to building docs"

gherceg committed:
Merge pull request #31575 from dimagi/revert-31571-gh/fix-docs-build

Revert "setup django app registry prior to building docs"

proteusvacuum committed:
Merge pull request #31443 from dimagi/fr/ucr-registry-linked

UCR Named Expression Registry

biyeun committed:
Update staging.yml

biyeun committed:
Update staging.yml

calellowitz committed:
Merge pull request #31574 from dimagi/ce/redis-api-throttle

only write to redis if throttling

esoergel committed:
Simplify method of accessing form meta

Not sure whether this was once used for something else, but at this
point this was essentially just a complicated way of getting `meta` out
fo `form_data`.

esoergel committed:
Set @xmlns to the one we use everywhere else

There are multiple xmlns attributes in xforms, but the one we want isn't
the one on the <meta /> node, but the non-namespaced one on the data
node.

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into fr/case-repeater-delete

proteusvacuum committed:
Merge pull request #31570 from dimagi/fr/case-repeater-delete

Custom repeater URLs with UCR expressions

Commit:
added aria label and tabindex to heading

gherceg committed:
Update translations.

[ci skip]

Robert-Costello committed:
Merge pull request #31576 from dimagi/rc/case-search-accessibility

Rc/case search accessibility

Robert-Costello committed:
Update staging.yml

gherceg committed:
add DOCS_BUILD flag to sphinx command to setup django

gherceg committed:
point to instructions in local.in

[ci skip]

akashkj commented:
@snopoke This is for superset instance where we want domains for which analytics flags is enabled, so added as optional argument.

snopoke reviewed (approved):
Seems OK I guess though it doesn't seem like a generally useful feature.

akashkj commented on corehq/apps/api/resources/v0_5.py:
```diff
@@ -860,15 +859,18 @@ def dispatch_list(self, request, **kwargs):
             if not domain_has_privilege(domain, privileges.ZAPIER_INTEGRATION):
                 continue
             domain_object = Domain.get_by_name(domain)
+            if feature_flag and feature_flag not in toggles.toggles_enabled_for_domain(domain):
```
Can do that, my thought was that checking 'flag in []' would implicitly handle the wrong flag since no domain's enabled flags will match it.
 Alternatively, if we should throw error then can make an explicit check for it. What do you suggest?

snopoke replied:
Returning a 400 Bad Request would be best practice. Not doing so also makes it impossible to tell if you got the name wrong or if the domain doesn't have the flag enabled.

Since these names are just python variable names too it's possible that someone could rename the toggle in python and break the integration.

gherceg committed:
add comment explaining why we need django.setup()

gherceg committed:
Merge pull request #31579 from dimagi/gh/update-readme

Point to instructions in local.in for installing requirements

gherceg committed:
Merge pull request #31578 from dimagi/gh/fix-docs

Add DOCS_BUILD flag to sphinx command to setup django before builds

esoergel committed:
Use normalized form.metadata

orangejenny committed:
Update staging.yml

orangejenny committed:
Update staging.yml

joxl committed:
Merge pull request #31125 from dimagi/jm/elastic-client-adapter

Elastic Client Adapter: initial implementation

gherceg committed:
Update staging.yml

proteusvacuum committed:
Keep date strings marked as constants as strings

orangejenny committed:
Removed unused "no forms" build error

proteusvacuum committed:
Keep converted dates; convert them to strings if they are ethiopian

mjriley committed:
Merge pull request #30683 from dimagi/mjr/report_end_dates

Use consistent date formatting for report filters

proteusvacuum committed:
Update tests - you can now convert native dates

joxl committed:
Update staging.yml

joxl committed:
Update staging.yml

gherceg committed:
Update staging.yml

gherceg committed:
Update staging.yml

snopoke committed:
test to validate case block ordering in a form does not impact index creation

snopoke committed:
check for any extension index when looking for case claim cases

snopoke committed:
test for deleted index on claim case

kaapstorm committed:
Drop corehq.util.django2_shim

django2_shim was created as a tuil to help upgrade from Django 1 to
Django 2. We don't need it any more.

snopoke committed:
add missing imports

snopoke committed:
isort

snopoke committed:
Merge pull request #31588 from dimagi/sk/add-missing-imports

add missing imports

snopoke committed:
Merge pull request #31586 from dimagi/sk/case-ordering-test

test to validate case block ordering in a form does not impact index creation

esoergel committed:
Merge pull request #31577 from dimagi/es/system-xmlns

Show proper XMLNS on form_data page

proteusvacuum committed:
Merge pull request #31582 from dimagi/fr/ethiopian-constants

Ethiopian Dates: Keep date strings marked as constants as strings

orangejenny committed:
Merge pull request #31583 from dimagi/orangejenny-patch-10

Removed unused "no forms" build error

gherceg committed:
update docs about docs

kaapstorm committed:
Merge pull request #31587 from dimagi/nh/django2_shim

Drop corehq.util.django2_shim

gherceg committed:
Merge pull request #31589 from dimagi/gh/update-docs-for-docs-about-docs-build

Add DOCS_BUILD info to docs/documenting.rst

gherceg committed:
modify exception and error message to make sense

gherceg committed:
consolidate into ImporterFileNotFound exception

gherceg committed:
run isort

gherceg committed:
update test assertion

snopoke committed:
Merge pull request #31585 from dimagi/sk/host-index-exists

Minor fixes for finding already claimed cases

snopoke committed:
Merge pull request #31494 from dimagi/sk/remove-feature-phone-support

[tech debt] remove feature phone support

Charl1996 committed:
updating files: scripts/staging.yml

[ci skip]

Charl1996 committed:
updating files: scripts/staging.yml

[ci skip]

Commit:
Merge branch 'aj/domain-membership-with-feature-flag' of https://github.com/dimagi/commcare-hq into aj/domain-membership-with-feature-flag

gherceg committed:
Merge pull request #31590 from dimagi/gh/update-error

Modify exception raised and error shown to user when case import cache issue encountered

Charl1996 committed:
updating files: scripts/staging.yml

[ci skip]

Charl1996 committed:
updating files: scripts/staging.yml

[ci skip]

orangejenny committed:
Added save button handler for new filter input

snopoke committed:
remove icds tagging of sms metrics

Robert-Costello committed:
Update staging.yml

snopoke committed:
check if owner_id is blank

snopoke committed:
remove unnecessary tuple braces

joxl committed:
Merge pull request #31595 from dimagi/jm/fix-alembic-test

Fix failing test (needed for alembic upgrade)

snopoke committed:
test cases

orangejenny committed:
Merge pull request #31565 from dimagi/jls/tab-nodeset-filter

Add filter to detail tab nodeset

snopoke committed:
Merge pull request #31598 from dimagi/sk/remove-icds-indiator

remove icds tagging of sms metrics

snopoke committed:
Merge pull request #31599 from dimagi/sk/empty-owner

don't 500 if owner_id is empty

Commit:
Refactor based on testing

Commit:
Merge branch 'aj/domain-membership-with-feature-flag' of https://github.com/dimagi/commcare-hq into aj/domain-membership-with-feature-flag
